library/core/src/ptr/mod.rs

```diff
@@ -2164,13 +2164,36 @@ pub fn addr_eq<T: ?Sized, U: ?Sized>(p: *const T, q: *const U) -> bool {
 
 /// Compares the *addresses* of the two function pointers for equality.
 ///
-/// Function pointers comparisons can have surprising results since
-/// they are never guaranteed to be unique and could vary between different
-/// code generation units. Furthermore, different functions could have the
-/// same address after being merged together.
+/// This is the same as `f == g`, but using this function makes clear that the potentially
+/// surprising semantics of function pointer comparison are involved.
+/// There are very few guarantees about how functions are compiled and they have no intrinsic
+/// “identity”; in particular, this comparison:
+///
+/// * May return `true` unexpectedly, in cases where functions are equivalent.
+///   For example, the following program is likely (but not guaranteed) to print `(true, true)`
+///   when compiled with optimization:
+///
+///   ```
+///   # #![feature(ptr_fn_addr_eq)]
+///   let f: fn(i32) -> i32 = |x| x;
+///   let g: fn(i32) -> i32 = |x| x + 0;  // different closure, different body
+///   let h: fn(u32) -> u32 = |x| x + 0;  // different signature too
+///   dbg!(std::ptr::fn_addr_eq(f, g), std::ptr::fn_addr_eq(f, h));
+///   ```
+///
+/// * May return `false` in any case.
+///   This is particularly likely with generic functions but may happen with any function.
+///   (From an implementation perspective, this is possible because functions may sometimes be
+///   processed more than once by the compiler, resulting in duplicate machine code.)
+///
+/// Despite these false positives and false negatives, this comparison can still be useful.
```

Comment:
true & false ? instead of false & false.
```suggestion
/// Despite these true positives and false negatives, this comparison can still be useful.
```

Reply:
Whether a result is a false positive or a true positive (or a false negative or a true negative) depends on what you want fn pointer comparison to do.
* If you want to compare the address the function pointer corresponds to, there should be no false (incorrect) results (positive or negative) IIUC (except maybe due to LLVM pointer comparison wierdness w.r.t. provenance?).
* If you want to see if two fn pointers correspond to the same function, there may be false positives (fn pointers you got from different functions comparing equal) and false negatives (fn pointers you got from the same function comparing inequal).

Given the above context describing the second viewpoint, I think "false positives and false negatives" is correct here.

Reply:
Oh, yeah never-mind. I read it as "false positives" two times.